Gradle 3.5 compatibility problem

[//]: # (
. Note: for support questions, please use Stackoverflow or Gitter**. 
. This repository's issues are reserved for feature requests and bug reports.
.
. In case of any problems with Allure Jenkins plugin** please use the following repository 
. to create an issue: https://github.com/jenkinsci/allure-plugin/issues
.
. Make sure you have a clear name for your issue. The name should start with a capital 
. letter and no dot is required in the end of the sentence. An example of good issue names:
.
. - The report is broken in IE11
. - Add an ability to disable default plugins
. - Support emoji in test descriptions
)

#### I'm submitting a ... 
  - [x] bug report
  - [ ] feature request
  - [ ] support request => Please do not submit support request here, see note at the top of this template.

#### What is the current behavior?

```
Caused by: java.lang.NoSuchMethodError: org.gradle.api.tasks.testing.Test.setJvmArgs(Ljava/util/List;)V
        at io.qameta.allure.gradle.AllurePlugin$_applyAspectjweaver_closure4$_closure6.doCall(AllurePlugin.groovy:132)
        at io.qameta.allure.gradle.AllurePlugin$_applyAspectjweaver_closure4$_closure6.call(AllurePlugin.groovy)
        at org.gradle.api.internal.AbstractTask$ClosureTaskAction.execute(AbstractTask.java:656)
        at org.gradle.api.internal.AbstractTask$ClosureTaskAction.execute(AbstractTask.java:637)
```

#### If the current behavior is a bug, please provide the steps to reproduce and if possible a minimal demo of the problem

```
allure {
   autoconfigure = true
   version = 2.1.1
}
```
`./gradlew build`


#### What is the expected behavior?

build passed

#### What is the motivation / use case for changing the behavior?

#### Please tell us about your environment:

- Allure version:        2.1.1
- Test framework:        any
- Allure adaptor:        any
- Generate report using: allure-gradle@2.2

#### Other information 

 - Gradle version:  3.5

[//]: # (
. e.g. detailed explanation, stacktraces, related issues, suggestions 
. how to fix, links for us to have more context, eg. Stackoverflow, Gitter etc
)

--- a/build.gradle
+++ b/build.gradle
@@ -16,7 +16,6 @@ buildscript {
 
 apply plugin: 'java'
 apply plugin: 'groovy'
-apply plugin: 'maven-publish'
 apply plugin: 'java-gradle-plugin'
 
 apply from: "${rootProject.projectDir}/gradle/plugin-publish.gradle"

--- a/build.gradle
+++ b/build.gradle
@@ -16,7 +16,7 @@ buildscript {
 
 apply plugin: 'java'
 apply plugin: 'groovy'
-apply plugin: 'maven'
+apply plugin: 'maven-publish'
 apply plugin: 'java-gradle-plugin'
 
 apply from: "${rootProject.projectDir}/gradle/plugin-publish.gradle"

--- a/src/main/groovy/io/qameta/allure/gradle/AllurePlugin.groovy
+++ b/src/main/groovy/io/qameta/allure/gradle/AllurePlugin.groovy
@@ -128,9 +128,8 @@ class AllurePlugin implements Plugin<Project> {
 
             project.tasks.withType(Test).each { test ->
                 test.doFirst {
-                    List<String> jvmArgsWithJavaAgent = new ArrayList<String>(test.jvmArgs)
-                    jvmArgsWithJavaAgent.add("-javaagent:${aspectjConfiguration.singleFile}".toString())
-                    test.jvmArgs = jvmArgsWithJavaAgent as Iterable<?>
+                    String javaAgent = "-javaagent:${aspectjConfiguration.singleFile}"
+                    test.jvmArgs = [javaAgent] + test.jvmArgs as Iterable
                 }
                 if (project.logger.debugEnabled) {
                     project.logger.debug "jvmArgs for task $test.name $test.jvmArgs"

--- a/src/main/groovy/io/qameta/allure/gradle/AllurePlugin.groovy
+++ b/src/main/groovy/io/qameta/allure/gradle/AllurePlugin.groovy
@@ -128,8 +128,9 @@ class AllurePlugin implements Plugin<Project> {
 
             project.tasks.withType(Test).each { test ->
                 test.doFirst {
-                    String javaAgent = "-javaagent:${aspectjConfiguration.singleFile}"
-                    test.jvmArgs = [javaAgent] + test.jvmArgs
+                    List<String> jvmArgsWithJavaAgent = new ArrayList<String>(test.jvmArgs)
+                    jvmArgsWithJavaAgent.add("-javaagent:${aspectjConfiguration.singleFile}".toString())
+                    test.jvmArgs = jvmArgsWithJavaAgent as Iterable<?>
                 }
                 if (project.logger.debugEnabled) {
                     project.logger.debug "jvmArgs for task $test.name $test.jvmArgs"

--- a/src/test/java/io/qameta/allure/gradle/DependenciesTest.java
+++ b/src/test/java/io/qameta/allure/gradle/DependenciesTest.java
@@ -22,22 +22,31 @@ import static org.gradle.testkit.runner.TaskOutcome.SUCCESS;
 @RunWith(Parameterized.class)
 public class DependenciesTest {
 
-    @Parameterized.Parameter
+    @Parameterized.Parameter(0)
+    public String version;
+
+    @Parameterized.Parameter(1)
     public String project;
 
     @Rule
     public GradleRunnerRule gradleRunner = new GradleRunnerRule()
+            .version(() -> version)
             .project(() -> project)
             .tasks("test");
 
-    @Parameterized.Parameters(name = "{0}")
+    @Parameterized.Parameters(name = "{1} [{0}]")
     public static Collection<Object[]> getFrameworks() {
         return Arrays.asList(
-                new Object[]{"src/it/cucumber-jvm"},
-                new Object[]{"src/it/junit4"},
-                new Object[]{"src/it/junit4-autoconfigure"},
-                new Object[]{"src/it/testng"},
-                new Object[]{"src/it/testng-autoconfigure"}
+                new Object[]{"3.5", "src/it/cucumber-jvm"},
+                new Object[]{"4.0", "src/it/cucumber-jvm"},
+                new Object[]{"3.5", "src/it/junit4"},
+                new Object[]{"4.0", "src/it/junit4"},
+                new Object[]{"3.5", "src/it/junit4-autoconfigure"},
+                new Object[]{"4.0", "src/it/junit4-autoconfigure"},
+                new Object[]{"3.5", "src/it/testng"},
+                new Object[]{"4.0", "src/it/testng"},
+                new Object[]{"3.5", "src/it/testng-autoconfigure"},
+                new Object[]{"4.0", "src/it/testng-autoconfigure"}
         );
     }
 

--- a/src/test/java/io/qameta/allure/gradle/ReportTest.java
+++ b/src/test/java/io/qameta/allure/gradle/ReportTest.java
@@ -18,22 +18,28 @@ import static org.gradle.testkit.runner.TaskOutcome.SUCCESS;
 public class ReportTest {
 
     @Parameterized.Parameter(0)
-    public String project;
+    public String version;
 
     @Parameterized.Parameter(1)
+    public String project;
+
+    @Parameterized.Parameter(2)
     public String[] tasks;
 
     @Rule
     public GradleRunnerRule gradleRunner = new GradleRunnerRule()
+            .version(() -> version)
             .project(() -> project)
             .tasks(() -> tasks);
 
-    @Parameterized.Parameters(name = "{0}")
+    @Parameterized.Parameters(name = "{1} [{0}]")
     public static Collection<Object[]> getFrameworks() {
         return Arrays.asList(
-                new Object[]{"src/it/test-finalized-by-report", new String[]{"test"}},
-                new Object[]{"src/it/report-task", new String[]{"allureReport"}},
-                new Object[]{"src/it/report-multi", new String[]{"allureAggregatedReport"}}
+                new Object[]{"3.5", "src/it/test-finalized-by-report", new String[]{"test"}},
+                new Object[]{"3.5", "src/it/report-task", new String[]{"allureReport"}},
+                new Object[]{"4.0", "src/it/report-task", new String[]{"allureReport"}},
+                new Object[]{"3.5", "src/it/report-multi", new String[]{"allureAggregatedReport"}},
+                new Object[]{"4.0", "src/it/report-multi", new String[]{"allureAggregatedReport"}}
         );
     }
 

--- a/src/test/java/io/qameta/allure/gradle/TestNgSpiOffTest.java
+++ b/src/test/java/io/qameta/allure/gradle/TestNgSpiOffTest.java
@@ -14,6 +14,7 @@ public class TestNgSpiOffTest {
 
     @Rule
     public GradleRunnerRule gradleRunner = new GradleRunnerRule()
+            .version("4.0")
             .project("src/it/testng-spi-off")
             .tasks("test");
 

--- a/src/test/java/io/qameta/allure/gradle/rule/GradleRunnerRule.java
+++ b/src/test/java/io/qameta/allure/gradle/rule/GradleRunnerRule.java
@@ -27,6 +27,8 @@ public class GradleRunnerRule extends ExternalResource {
 
     private Supplier<String[]> tasksSupplier;
 
+    private Supplier<String> versionSupplier;
+
     private BuildResult buildResult;
 
     private File projectDir;
@@ -37,6 +39,11 @@ public class GradleRunnerRule extends ExternalResource {
 
     public File getProjectDir() {
         return this.projectDir;
+    }
+
+    public GradleRunnerRule version(Supplier<String> version) {
+        this.versionSupplier = version;
+        return this;
     }
 
     public GradleRunnerRule project(Supplier<String> supplier) {
@@ -63,6 +70,7 @@ public class GradleRunnerRule extends ExternalResource {
         buildResult = GradleRunner.create()
                 .withProjectDir(projectDir)
                 .withArguments(tasksSupplier.get())
+                .withGradleVersion(versionSupplier.get())
                 .withTestKitDir(new File(projectDir.getParentFile().getAbsolutePath(), ".gradle"))
                 .withPluginClasspath(readPluginClasspath())
                 .build();

--- a/src/test/java/io/qameta/allure/gradle/rule/GradleRunnerRule.java
+++ b/src/test/java/io/qameta/allure/gradle/rule/GradleRunnerRule.java
@@ -41,6 +41,10 @@ public class GradleRunnerRule extends ExternalResource {
         return this.projectDir;
     }
 
+    public GradleRunnerRule version(String version) {
+        return version(() -> version);
+    }
+
     public GradleRunnerRule version(Supplier<String> version) {
         this.versionSupplier = version;
         return this;

